hedera-node/hedera-app/src/test/java/com/hedera/node/app/util/HashTreeManagerTest.java

```diff
@@ -0,0 +1,168 @@
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.util;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.hedera.hapi.block.stream.Block;
+import com.hedera.hapi.block.stream.BlockHeader;
+import com.hedera.hapi.block.stream.BlockItem;
+import com.hedera.hapi.block.stream.EventMetadata;
+import com.hedera.hapi.block.stream.output.StateChanges;
+import com.hedera.hapi.block.stream.output.TransactionOutput;
+import com.hedera.hapi.block.stream.output.TransactionResult;
+import com.hedera.hapi.node.base.Transaction;
+import com.hedera.pbj.runtime.Codec;
+import com.hedera.pbj.runtime.ParseException;
+import com.hedera.pbj.runtime.io.ReadableSequentialData;
+import com.hedera.pbj.runtime.io.WritableSequentialData;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import com.swirlds.common.merkle.MerkleNode;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.io.IOException;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.Test;
+
+public class HashTreeManagerTest {
+
+    private final BlockItem blockItem = BlockItem.newBuilder().build();
+    private final Block block = Block.newBuilder().items().build();
+    private final Map<BlockHeader, MerkleNode> inputRootsByBlock = new HashMap<>();
+    private final Map<BlockHeader, MerkleNode> outputRootsByBlock = new HashMap<>();
+    private final String SHA_384_ROOT_HASH =
+            "1333c0960fbb9cac040834be1597b45571dae1bc14b3f78281c20dd147826f017ce4f4105d71ecf394bb41db424a637a";
+
+    @Test
+    public void testHashTreeManager() throws NoSuchAlgorithmException {
+        HashTreeManager<String> tree = new HashTreeManager<>(new SimpleStringCodec());
+        List<String> elements = Arrays.asList("Event1", "Transaction1", "Transaction2", "Event2", "Transaction3");
+        tree.addElements(elements);
+
+        // Calculate the root hash
+        Bytes rootHash = tree.getTreeRoot();
+        assertThat(rootHash).isNotNull();
+        assertThat(rootHash.toString()).isEqualTo(SHA_384_ROOT_HASH);
+    }
+
+    @Test
+    public void testIncrementalConstruction() throws Exception {
+        HashTreeManager<String> tree = new HashTreeManager<>(new SimpleStringCodec());
+
+        tree.addElement("leaf1");
+        // A Merkle tree with a single leaf does have a root hash.
+        assertThat(tree.getTreeRoot()).isNotNull();
+
+        tree.addElement("leaf2");
+        String rootHash = tree.getTreeRootAsString();
+        assertThat(rootHash)
+                .isEqualTo(tree.getTreeRootAsString()); // Root hash should remain consistent after balancing
+
+        tree.addElement("leaf3");
+        assertThat(tree.getTreeRoot()).isNotNull(); // A tree with three leaves still computes a root hash
```

Comment:
I like the idea of this test, but wonder if we need more in the tree's public API in order to verify correct ordering. As it stands, really we're just testing that there's a root node. We're not checking that the hash is what we expect it to be, which means we could be adding them in any order and this test would only verify that `getTreeRootAsString()` returns not null. (Especially since the implementation isn't a tree at all, just an ordered collection of hashes!) The ordering is crucial if a block node wants to verify, so I think the order should be tested. 

Part of this ordering is testing the exact value of the hashes instead of just checking for non-null; the hashes are computed by concatenating child hashes together, so if they're in the wrong order or the tree isn't balanced correctly, the hash will be different from the expected. 

The other part of ordering is perhaps verifying the hashes of the "intermediate" parent nodes, though I suppose one could argue that those will also be reflected in the hash values if they're incorrect. Thoughts?

Reply:
I've added a new method `getHashList() `to `HashTreeManager` and used that to get and verify the hashes in the test at each point after adding leaves to the tree. 

This gives us coverage to validate the correctness of the ordering without exposing too much of the tree's implementation. 

As discussed in slack we may want to revisit once we have integrated more into `HandleWorkflow` however and add the ability to verify for example by passing in a path to the node and receiving the subroot/node, e.g. `root -> left -> right -> left -> left`